### ISSUE TITLE
Revamp Review Workflow

### ISSUE THREAD

**Issue by pauljamescleary** [vinyldns/vinyldns#839]
## Background
The manual review workflow that is currently implemented is hard to follow across the various states and transitions.  The purpose of this effort is to streamline the workflow, and finalize and reshape the states and workflows to be in a better place.

## Proposed State Machine

![Proposed State Machine](http://www.plantuml.com/plantuml/png/TP1DJiCm48NtFiKi6r0lWAgeLPGk488eGMmG2nDFYOL_8dj2nEsnLLLsZRhClE_BFC_U4tUqQCKcAH1Gp_GB0ltev4pI6iO2G0Z3ygab4GhOxKxMVTSrUx_xWEssJX5BK9ntQ8GqNpt-I5pX0JOhToOCFCmEFHggIi7P4RqFIAIvKZ0gv0wC9V3ZX69HA4ffk0JNpdrkiEigKGlbWz0SdflswMRSjuzjyrh6PTD3r_KlRoriZ8NB4wj2Afec9xKiZVA2kRLAch9TBFtwNxjWDEqmf4s0Yi_-V9O4Hk_8PNR7K_XQ7Fggn1ghPuM44JJsDjTnHvAhupcFjBSQCtwFHYnQ_G40)

<details>
<summary>diagram source</summary>
<p>

```
@startuml
hide empty description

state Submitted <<fork>>
[*] --> Submitted
Submitted --> PendingReview : (warnings present)
Submitted --> Processing : (all clear not scheduled)
Submitted --> Scheduled : (all clear + scheduled)

PendingReview --> [*] : CANCEL
PendingReview --> [*] : REJECT
PendingReview --> Processing : APPROVE (no warnings)
PendingReview --> Scheduled : APPROVE (scheduled)
PendingReview --> PendingReview : APPROVE (with warnings)

Scheduled --> Processing : PROCESS (schedule elapsed)

Processing --> Failed : (all failures)
Processing --> Complete : (no failures)
Processing --> PartialFailure : (some failures)
@enduml
```

**Comment by britneywright** [comment]
Since we don't have Scheduled batch changes processing automatically should the `Validate` path be for both validate and approval?

**Comment by pauljamescleary** [comment]
@britneywright Right, I think `VALIDATE` is just "check if this is ok".  It makes sense in a `Scheduled` state because technicians can double check that off hours scheduled DNS changes will be good to go before leaving work.

I think `APPROVE` implies validation.  The intention of `APPROVE` is to advance the change to it's logical next state (Processing or Scheduled).

Does that answer your question?

**Comment by mitruly** [comment]
@pauljamescleary Not sure if @britneywright is mentioning that she would like to see a `VALIDATE (no new warnings)` for `Scheduled` which I think makes sense.

**Comment by britneywright** [comment]
The list of scheduled actions should include `CANCEL` and `REJECT`

**Comment by mitruly** [comment]
@britneywright There is an `ABANDON (reject or cancel)` path

**Comment by britneywright** [comment]
@mitruly In the diagram yes, but i'm talking about the list of actions in the workflow description text. The only actions listed under Scheduled currently are `PROCESS` and `VALIDATE`.

**Comment by pauljamescleary** [comment]
@britneywright, @mitruly right, I updated the paths to just `ABANDON` to make the diagram less busy.  The description includes `CANCEL` and `REJECT`

I also added a path on `VALIDATE` from `Scheduled --> Scheduled` in the event that no warnings remain.

**Comment by pauljamescleary** [comment]
@britneywright we should not allow `REJECT` in a `Scheduled` state.  That is only available in `PendingReview`.

**Comment by pauljamescleary** [comment]
@britneywright updated to put REJECT back in.

**Comment by pauljamescleary** [comment]
I believe that we need to change the way we generate the batch change status and really make it simpler than it is now, where it all lives inside the `BatchChange calculateBatchStatus` ...

The primary change is to save the batch change in the database.  This involves the following:

1. Add a column / formal attribute of the batch change named `status`
2. Status should be deterministically set based on processing.  The following highlight the changes
    1. If REJECTED successfully --> set status to `Rejected` before saving
    1. If CANCELLED successfully --> set status to `Cancelled` before saving
    1. If automatically approved (no warnings) --> set status to `Scheduled` if it is scheduled
    1. If manually approved --> set status to `Scheduled` if it is scheduled
    1. If there are any warnings present --> set status to `PendingReview`

I believe by in large everything kinda remains the same.

**Comment by britneywright** [comment]
don't forget the existing statuses, Complete, Failed and PartialFailure that will come after PendingReview or Scheduled. 

This kind of brings us back to approval status and how the two are essentially the same except that an approved batch change could fail or partial fail during the backend processing
